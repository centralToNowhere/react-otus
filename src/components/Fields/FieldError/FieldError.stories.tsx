import React from "react";
import { Story, Meta } from "@storybook/react";
import { FieldError, IFieldErrorProps } from "./FieldError";

export default {
  title: "Components/FieldError",
  component: FieldError,
  args: {
    show: true,
    msg: "something went wrong",
  },
} as Meta;

const Template: Story<IFieldErrorProps> = (args) => <FieldError {...args} />;

export const ShowError = Template.bind({});

ShowError.args = {
  show: true,
};

export const NoError = Template.bind({});

NoError.args = {
  show: false,
};
