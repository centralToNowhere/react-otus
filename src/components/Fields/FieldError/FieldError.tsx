import React from "react";
import styled from "@emotion/styled";

export interface IFieldErrorProps {
  show: boolean;
  msg: string;
}

export const initialErrorProps = {
  show: false,
  msg: "",
};

export const mergeErrorMessages = (
  errors: Array<{
    show: boolean;
    msg: string;
  }>
): string => {
  return errors.reduce((msg, error) => {
    return error.show ? `${msg}${error.msg}\n` : msg;
  }, "");
};

export const FieldError: React.FC<IFieldErrorProps> = ({
  show = false,
  msg = "",
}: IFieldErrorProps) => {
  return (
    <>
      {show && (
        <ErrorContainer>
          <span role="alert" className="field-error-msg">
            {msg}
          </span>
        </ErrorContainer>
      )}
    </>
  );
};

const ErrorContainer = styled.div`
  position: relative;

  .field-error-msg {
    white-space: pre-wrap;
    color: red;
    position: absolute;
  }
`;
