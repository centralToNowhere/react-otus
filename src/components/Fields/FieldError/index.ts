export { FieldError } from "./FieldError";
export type { IFieldErrorProps } from "./FieldError";
