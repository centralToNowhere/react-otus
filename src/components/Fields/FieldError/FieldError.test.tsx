import React from "react";
import { render, screen } from "@testing-library/react";
import { FieldError } from "./FieldError";

describe("error field tests", () => {
  it("should show error", () => {
    const { asFragment } = render(
      <FieldError show={true} msg={"something went wrong"} />
    );

    const errorContainer: HTMLDivElement = screen.getByText(
      "something went wrong"
    );

    expect(errorContainer).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not show error", () => {
    render(<FieldError show={false} msg={"something went wrong"} />);

    const errorContainer: HTMLDivElement | null = screen.queryByText(
      "something went wrong"
    );

    expect(errorContainer).not.toBeInTheDocument();
  });
});
