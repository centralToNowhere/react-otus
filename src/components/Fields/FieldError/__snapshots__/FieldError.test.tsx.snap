// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error field tests should show error 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
}

.emotion-0 p.error-msg {
  position: absolute;
  line-height: 1em;
  font-size: 1em;
  color: red;
}

<div
    class="emotion-0 emotion-1"
  >
    <p
      class="error-msg"
    >
      something went wrong
    </p>
  </div>
</DocumentFragment>
`;
